src/EncryptStreamWrapper.php

```diff
@@ -0,0 +1,193 @@
+   * @param string $encryption_profile
+   *
+   * @return string
+   */
+  protected function encrypt($raw_file, $encryption_profile) {
+    /** @var \Drupal\encrypt\EncryptService $encryption */
+    $encryption = \Drupal::service('encryption');
+    return $encryption->encrypt($raw_file, $this->getEncryptionProfile($encryption_profile));
+  }
+
+  /**
+   * @param string $encryption_profile
+   *
+   * @return \Drupal\Core\Entity\EntityInterface|\Drupal\encrypt\EncryptionProfileInterface|null
+   */
+  protected function getEncryptionProfile($encryption_profile) {
+    /** @var \Drupal\encrypt\EncryptionProfileManager $profile_manager */
+    $profile_manager = \Drupal::service('encrypt.encryption_profile.manager');
+    return $profile_manager->getEncryptionProfile($encryption_profile);
+  }
+
+  protected function extractEncryptionProfile($uri) {
+    return $this->encryptionProfile = parse_url($uri, PHP_URL_HOST);
+  }
+  /**
+   * {@inheritdoc}
+   */
+  public function stream_open($uri, $mode, $options, &$opened_path) {
+    // Create Encrypted Files directory if it doesn't exist.
+    $fe_directory = $this->getDirectoryPath();
+    if ($fe_directory && !file_exists($fe_directory)) {
+      mkdir($fe_directory, 0755);
+    }
+
+    $this->extractEncryptionProfile($uri);
+
+    // Load resource location.
+    $this->uri = $uri;
+    $path = $this->getLocalPath();
+    // Save the mode for later reference.
+    $this->mode = $mode;
+    // Load temp file as our handle.
+    $this->handle = fopen('php://memory', 'w+b');
+    // If file exists, decrypt and load it into memory.
+    if (file_exists($path)) {
+      $raw_file = file_get_contents($path);
+      $decrypted_file = $this->decrypt($raw_file, $this->encryptionProfile);
+      $this->setFileInfo($decrypted_file, $uri);
+      // Write to memory.
+      fwrite($this->handle, $decrypted_file);
+      rewind($this->handle);
+    }
+    // Set $opened_path.
+    if ((bool) $this->handle && $options & STREAM_USE_PATH) {
+      $opened_path = $path;
+    }
+
+    return (bool) $this->handle;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  protected function getTarget($uri = NULL) {
+    if (!isset($uri)) {
+      $uri = $this->uri;
+    }
+
+    $target = parse_url($uri, PHP_URL_PATH);
+
+    // Remove erroneous leading or trailing, forward-slashes and backslashes.
+    return trim($target, '\/');
+  }
+
+  /**
+   * Encrypts and writes the open file to disk, then closes the stream.
+   */
+  public function stream_close() {
+    // If file mode we opened with is only for reading,
+    // don't resave the file.
+    if ((strpos($this->mode, 'r') !== FALSE) &&
+      (strpos($this->mode, '+') === FALSE)) {
+      fclose($this->handle);
+      return;
+    }
+    // Encrypt file and save.
+    rewind($this->handle);
+    $file_contents = stream_get_contents($this->handle);
+    file_put_contents($this->getLocalPath(), $this->encrypt($file_contents, $this->encryptionProfile));
```

Comment:
Just wondering, this seems to suggest this writes out the encrypted file upon close. Can it/should it write before hand using the streaming algorithm?